cloud_watch_block.py

```diff
@@ -0,0 +1,146 @@
+from nio.metadata.properties import ObjectProperty, PropertyHolder, \
+    IntProperty, StringProperty, SelectProperty
+from nio.metadata.properties.version import VersionProperty
+from nio.modules.threading import Lock
+
+from boto.ec2.cloudwatch import connect_to_region
+
+
+class AWSRegion(Enum):
+    us_east_1 = 0
+    us_west_2 = 1
+    eu_west_1 = 2
+
+
+class MetricStatistics(Enum):
+    Average = 0
+    Sum = 1
+    SampleCount = 2
+    Maximum = 3
+    Minimum = 4
+
+
+class AWSCreds(PropertyHolder):
+    access_key = StringProperty(title="Access Key",
+                                default="[[AMAZON_ACCESS_KEY_ID]]")
+    access_secret = StringProperty(title="Access Secret",
+                                   default="[[AMAZON_SECRET_ACCESS_KEY]]")
+    region = SelectProperty(
+        AWSRegion, default=AWSRegion.us_east_1, title="AWS Region")
+
+
+@Discoverable(DiscoverableType.block)
+class CloudWatch(Block):
+
+    creds = ObjectProperty(AWSCreds, title="AWS Credentials")
+    metric = StringProperty(
+        title="Metric to Monitor", default="CPUCreditBalance")
+    lookback_mins = IntProperty(title="Lookback Minutes", default=5)
+    result_period = IntProperty(title="Result Period Minutes", default=5)
+    statistic = SelectProperty(MetricStatistics, title="Statistic Type",
+                               default=MetricStatistics.Average)
+    version = VersionProperty('0.0.1')
+
+    def __init__(self):
+        super().__init__()
+        self._conn = None
+        self._metrics = []
+        self._metrics_lock = Lock()
+
+    def configure(self, context):
+        super().configure(context)
+        # Set boto's log level to be the same as ours
+        logging.getLogger('boto').setLevel(self._logger.logger.level)
+        self._connect()
+        self._sync_metrics()
+
+    def _connect(self):
+        """ Create a connection to AWS using our credentials """
+        region_name = re.sub('_', '-', self.creds.region.name)
+        self._logger.debug("Connecting to region {}...".format(region_name))
+        self._conn = connect_to_region(
+            region_name,
+            aws_access_key_id=self.creds.access_key,
+            aws_secret_access_key=self.creds.access_secret)
+        self._logger.debug("Connection complete")
+
+    def _sync_metrics(self):
+        """ Makes a request to CloudWatch and stores valid metrics
+
+        This method takes the configured metric name and obtains all of the
+        valid metrics that match that metric name. Metric objects contain
+        namespace, dimensions, and possibly some other important information
+        for querying later.
+
+        Returns:
+            None: saves the metics in the self._metrics list instead
+        """
+        with self._metrics_lock:
+            self._logger.debug(
+                "Syncing valid metrics for {} from EC2".format(self.metric))
+            metrics_res = self._conn.list_metrics(metric_name=self.metric)
```

Comment:
If credentials are invalid:


Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/boto/ec2/cloudwatch/__init__.py", line 286, in list_metrics
    return self.get_list('ListMetrics', params, [('member', Metric)])
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/boto/connection.py", line 1182, in get_list
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.BotoServerError: BotoServerError: 403 Forbidden
<ErrorResponse xmlns="http://monitoring.amazonaws.com/doc/2010-08-01/">
  <Error>
    <Type>Sender</Type>
    <Code>InvalidClientTokenId</Code>
    <Message>The security token included in the request is invalid.</Message>
  </Error>
  <RequestId>37425fea-e208-11e4-ac8a-9dde9c0aa0ed</RequestId>
</ErrorResponse>